data-prepper-plugins/common/src/main/java/com/amazon/dataprepper/plugins/buffer/BlockingBuffer.java

```diff
@@ -112,12 +113,17 @@ public void doWrite(T record, int timeoutInMillis) throws TimeoutException {
         final Stopwatch stopwatch = Stopwatch.createStarted();
         try {
             while (stopwatch.elapsed(TimeUnit.MILLISECONDS) < timeoutInMillis && records.size() < batchSize) {
-                final T record = blockingQueue.poll(timeoutInMillis, TimeUnit.MILLISECONDS);
-                if (record != null) { //record can be null, avoiding adding nulls
-                    records.add(record);
-                }
-                if (records.size() < batchSize) {
-                    blockingQueue.drainTo(records, batchSize - records.size());
+                synchronized (this) {
+                    final T record = blockingQueue.poll(timeoutInMillis, TimeUnit.MILLISECONDS);
+                    if (record != null) { //record can be null, avoiding adding nulls
+                        records.add(record);
+                    }
+                    numRecordsToBeChecked.incrementAndGet();
```

Comment:
This should be inside the `if(records!=null)`. Correct?

Reply:
+1